🌐 Add Portuguese translation for `docs/tutorial/path-params.md` (#3664)
fastapi/fastapi: docs/pt/docs/tutorial/path-params.md

> # Parâmetros do caminho da Url

Reviewer: Are we using "caminho" instead of "rota"?

Author: yes, we are. 
As a programmer I prefer to say "Path" instead of caminho, but as a newbie, I'd prefer to read on docs the word "caminho" as a translate of path, not "rota".

Reviewer: That's not what I meant.

We make choices about ambiguous or multiple possible terms.

It looks like we've chosen "rota" in the past, as you can see here: https://fastapi.tiangolo.com/pt/

Author: Hi Kludex, sorry about that. BTW I've changed all the words from "caminho" to "rota" accordingly.

As this is my first PR ever, I'm a little confused, should I make a new pr?

Reviewer: > As this is my first PR ever, I'm a little confused, should I make a new pr?

For what? There's no need to open a new PR.

Reviewer: >  BTW I've changed all the words from "caminho" to "rota" accordingly.

Cool!

> # Parâmetros do caminho da Url

Reviewer: ```suggestion
# Parâmetros do caminho da URL
```

Você pode declarar os "parâmetros" ou "variáveis" com a mesma sintaxe utilizada pelo formato de strings do Python:

```Python hl_lines="6-7"
{!../../../docs_src/path_params/tutorial001.py!}
```

O valor do parâmetro que foi passado ao path `item_id` será passado para a sua função como o argumento `item_id`.

Então, se você rodar este exemplo e for até <a href="http://127.0.0.1:8000/items/foo" class="external-link" target="_blank">http://127.0.0.1:8000/items/foo</a>, você verá a seguinte resposta:

```JSON
{"item_id":"foo"}
```

## Parâmetros de caminho com tipos

Você pode declarar o tipo de um parâmetro na função usando as anotações padrões do Python:

```Python hl_lines="7"
{!../../../docs_src/path_params/tutorial002.py!}
```

Nesse caso, `item_id` está sendo declarado como um `int`.

> !!! Check

Reviewer: I'm not sure what word are we using in Portuguese, but probably is not "Check". Maybe some page has it?
```suggestion
!!! Check "Something"
```

Reviewer: O "check" tava certo. O que eu quis dizer é que da pra mudar a mensagem que aparece. Se você colocar `!!! check Algo`, a palavra "Algo" aparece na caixinha. Só que eu não lembro se a gente já usou essa palavra...

Author: Perfeito, entendi e fiz a alteração de acordo.

                "item_id"
            ],
            "msg": "value is not a valid integer",
            "type": "type_error.integer"
        }
    ]
}
```

devido ao parâmetro da rota `item_id` ter um valor `"foo"`, que não é um `int`.

O mesmo erro apareceria se você tivesse fornecido um `float` ao invés de um `int`, como em: <a href="http://127.0.0.1:8000/items/4.2" class="external-link" target="_blank">http://127.0.0.1:8000/items/4.2</a>

!!! Verifique
	Então, com a mesma declaração de tipo do Python, o **FastAPI** dá pra você validação de dados.

	Observe que o erro também mostra claramente o ponto exato onde a validação não passou.

	Isso é incrivelmente útil enquanto se desenvolve e debuga o código que interage com a sua API.

## Documentação

Quando você abrir o seu navegador em <a href="http://127.0.0.1:8000/docs" class="external-link" target="_blank">http://127.0.0.1:8000/docs</a>, você verá de forma automática e interativa a documtação da API como:

<img src="/img/tutorial/path-params/image01.png">

!!! check
	Novamente, apenas com a mesma declaração de tipo do Python, o **FastAPI** te dá de forma automática e interativa a documentação (integrada com o Swagger UI).

> 	Veja que o parâmetro de caminho está declarado como sendo um inteiro (int).

Reviewer 2: ```suggestion
	Veja que o parâmetro de rota está declarado como sendo um inteiro (int).
```

## Beneficios baseados em padrões, documentação alternativa

Devido ao schema gerado ser o padrão do <a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md" class="external-link" target="_blank">OpenAPI</a>, existem muitas ferramentas compatíveis.

Por esse motivo, o próprio **FastAPI** fornece uma API alternativa para documentação (utilizando ReDoc), que você pode acessar em <a href="http://127.0.0.1:8000/redoc" class="external-link" target="_blank">http://127.0.0.1:8000/redoc</a>:

<img src="/img/tutorial/path-params/image02.png">

Da mesma forma, existem muitas ferramentas compatíveis. Incluindo ferramentas de geração de código para muitas linguagens.

## Pydantic

> Toda a validação de dados é feita por baixo dos panos pelo <a href="https://pydantic-docs.helpmanual.io/" class="external-link" target="_blank">Pydantic</a>, então você tem todos os beneficios disso. E assim você sabe que está em boas mãos.

Reviewer 3: ```suggestion
Toda a validação de dados é feita por baixo dos panos pelo <a href="https://pydantic-docs.helpmanual.io/" class="external-link" target="_blank">Pydantic</a>, então você tem todos os benefícios disso. E assim você sabe que está em boas mãos.
```

Você pode usar as mesmas declarações de tipo com `str`, `float`, `bool` e muitos outros tipos complexos de dados.

Vamos explorar muitos destes tipos nos próximos capítulos do tutorial.

## A ordem importa

> Quando você cria operações de path, você pode se deparar com situações onde você pode ter um path fixo.

Reviewer 3: ```suggestion
Quando você cria operações de caminho, você pode se deparar com situações onde você pode ter um caminho fixo.
```

> Algo como `/users/me` por exemplo, digamos que esse path seja utilizado para pegar dados sobre o usuário atual.

Reviewer 3: ```suggestion
Algo como `/users/me` por exemplo, digamos que esse caminho seja utilizado para pegar dados sobre o usuário atual.
```

> E então você pode ter também um path `/users/{user_id}` para pegar dados sobre um usuário específico associado a um ID de usuário.

Reviewer 3: ```suggestion
E então você pode ter também um caminho `/users/{user_id}` para pegar dados sobre um usuário específico associado a um ID de usuário.
```

> Porque as operações de path são avaliadas em ordem, você precisa ter certeza que o path para `/users/me` está sendo declarado antes do path `/users/{user_id}`:

Reviewer 3: ```suggestion
Porque as operações de caminho são avaliadas em ordem, você precisa ter certeza que o caminho para `/users/me` está sendo declarado antes do caminho `/users/{user_id}`:
```

> Porque as operações de rota são avaliadas em ordem, você precisa ter certeza que a rota para `/users/me` está sendo declarado antes do caminho `/users/{user_id}`:

Reviewer 2: ```suggestion
Porque as operações de rota são avaliadas em ordem, você precisa ter certeza que a rota para `/users/me` está sendo declarado antes da rota `/users/{user_id}`:
```

```Python hl_lines="6  11"
{!../../../docs_src/path_params/tutorial003.py!}
```

> Caso contrário, o path para `/users/{user_id}` coincidiria também para `/users/me`, "pensando" que estaria recebendo o parâmetro `user_id` com o valor de `"me"`.

Reviewer 3: ```suggestion
Caso contrário, o caminho para `/users/{user_id}` coincidiria também para `/users/me`, "pensando" que estaria recebendo o parâmetro `user_id` com o valor de `"me"`.
```

## Valores predefinidos

> Se você tem uma operação de path que recebe um parâmetro de path, mas que você queira que esses valores possíveis do parametro de path sejam pré-definidos, você pode usar <abbr title="Enumeration">`Enum`</abbr> padrão do Python.

Reviewer 3: ```suggestion
Se você tem uma operação de caminho que recebe um parâmetro de caminho, mas que você queira que esses valores possíveis do parâmetro de caminho sejam predefinidos, você pode usar <abbr title="Enumeration">`Enum`</abbr> padrão do Python.
```

### Criando uma classe `Enum`

Importe `Enum` e crie uma sub-classe que herde de `str` e de `Enum`.

Por herdar de `str` a documentação da API vai ser capaz de saber que os valores devem ser do tipo `string` e assim ser capaz de mostrar eles corretamente.

Assim, crie atributos de classe com valores fixos, que serão os valores válidos disponíveis.

```Python hl_lines="1  6-9"
{!../../../docs_src/path_params/tutorial005.py!}
```

!!! informação
    <a href="https://docs.python.org/3/library/enum.html" class="external-link" target="_blank">Enumerations (ou enums) estão disponíveis no Python</a> desde a versão 3.4.

!!! dica
	Se você está se perguntando, "AlexNet", "ResNet", e "LeNet" são apenas nomes de <abbr title="técnicamente, modelos de arquitetura de Deep Learning">modelos</abbr> de Machine Learning (aprendizado de máquina).

> ### Declare um *parâmetro de path*

Reviewer 3: ```suggestion
### Declare um *parâmetro de caminho*
```

> Logo, crie um *parâmetro de path* com anotações de tipo usando a classe enum que você criou (`ModelName`):

Reviewer 3: ```suggestion
Logo, crie um *parâmetro de caminho* com anotações de tipo usando a classe enum que você criou (`ModelName`):
```

```Python hl_lines="16"
{!../../../docs_src/path_params/tutorial005.py!}
```

### Revise a documentação

> Visto que os valores disponíveis para o parâmetro do path estão pré-definidos, a documentação interativa pode mostrar esses valores de uma forma bem legal:

Reviewer 3: ```suggestion
Visto que os valores disponíveis para o parâmetro do caminho estão predefinidos, a documentação interativa pode mostrar esses valores de uma forma bem legal:
```

<img src="/img/tutorial/path-params/image03.png">

### Trabalhando com os *enumeration* do Python

O valor do *parâmetro do path* será um *membro de enumeration*.

#### Compare *membros de enumeration*

Você pode comparar eles com o *membro de enumeration* no enum `ModelName` que você criou:

```Python hl_lines="17"
{!../../../docs_src/path_params/tutorial005.py!}
```

#### Obtenha o *valor de enumerate*

Você pode ter o valor exato de enumerate (um `str` nesse caso) usando `model_name.value`, ou em geral, `your_enum_member.value`:

```Python hl_lines="20"
{!../../../docs_src/path_params/tutorial005.py!}
```

!!! conselho
	Você também poderia acessar o valor `"lenet"` com `ModelName.lenet.value`

#### Retorne *membros de enumeration*

> Você pode retornar *membros de enum* do seu *path de operação*, em JSON body aninhado (por exemplo um `dict`).

Reviewer 3: ```suggestion
Você pode retornar *membros de enum* do seu *caminho de operação*, em um corpo JSON aninhado (por exemplo um `dict`).
```

> Eles serão convertidos para o seus valores correspondentes (strings nesse caso) antes de serem retornados ao client:

Reviewer 3: ```suggestion
Eles serão convertidos para o seus valores correspondentes (strings nesse caso) antes de serem retornados ao cliente:
```

```Python hl_lines="18  21  23"
{!../../../docs_src/path_params/tutorial005.py!}
```

> No seu client você vai obter uma resposta JSON como:

Reviewer 3: ```suggestion
No seu cliente você vai obter uma resposta JSON como:
```

```JSON
{
  "model_name": "alexnet",
  "message": "Deep Learning FTW!"
}
```

> ## Parâmetros de path que contém paths

Reviewer 3: ```suggestion
## Parâmetros de caminho que contém caminhos
```

> Digamos que você tenha uma *operação de path* com um path `/files/{file_path}`.

Reviewer 3: ```suggestion
Digamos que você tenha uma *operação de caminho* com um caminho `/files/{file_path}`.
```

> Mas você precisa que o próprio `file_path` contenha um *path*, como `home/johndoe/myfile.txt`.

Reviewer 3: ```suggestion
Mas você precisa que o próprio `file_path` contenha um *caminho*, como `home/johndoe/myfile.txt`.
```

Então, a URL para este arquivo deveria ser algo como: `/files/home/johndoe/myfile.txt`.

### Suporte do OpenAPI

> O OpenAPI não suporta uma maneira de declarar um *parâmetro de path* que contenha um *path* dentro, dado que isso poderia levar a cenários que são difíceis de testar e definir.

Reviewer 3: ```suggestion
O OpenAPI não suporta uma maneira de declarar um *parâmetro de caminho* que contenha um *caminho* dentro, dado que isso poderia levar a cenários que são difíceis de testar e definir.
```

No entanto, você pode fazer isso no **FastAPI**, usando uma das ferramentas internas do Starlette. 

> A documentação continuaria funcionando, ainda que não adicionaria nenhuma informação dizendo que o parâmetro deveria conter um path.

Reviewer 3: ```suggestion
A documentação continuaria funcionando, ainda que não adicionaria nenhuma informação dizendo que o parâmetro deveria conter um caminho.
```

> ### Conversor de path

Reviewer 3: ```suggestion
### Conversor de caminho
```

> Usando uma opção direta do Starlette você pode declarar um *parâmetro de path* contendo um *path* usando uma URL como:

Reviewer 3: ```suggestion
Usando uma opção direta do Starlette você pode declarar um *parâmetro de caminho* contendo um *caminho* usando uma URL como:
```

```
/files/{file_path:path}
```

> Nesse caso, o nome do parâmetro é `file_path`, e a última parte, `:path`, diz que o parâmetro deveria coincidir com qualquer *path*.

Reviewer 3: ```suggestion
Nesse caso, o nome do parâmetro é `file_path`, e a última parte, `:path`, diz que o parâmetro deveria coincidir com qualquer *caminho*.
```
